debezium-core/src/main/java/io/debezium/connector/common/BaseSourceTask.java

```diff
@@ -86,6 +90,23 @@ public final void start(Map<String, String> props) {
      */
     protected abstract ChangeEventSourceCoordinator start(Configuration config);
 
+    @Override
+    public final List<SourceRecord> poll() throws InterruptedException {
+        try {
+            return doPoll();
+        }
+        catch (RetriableException e) {
+            stop();
+            start(config);
+            throw e;
+        }
+    }
+
+    /**
+     * Returns the next batch of source records, if any are available.
+     */
+    public abstract List<SourceRecord> doPoll() throws InterruptedException;
+
     @Override
     public void commitRecord(SourceRecord record) throws InterruptedException {
         Map<String, ?> currentOffset = record.sourceOffset();
```

Comment:
`commit` and `commitRecord` are called asynchrnously by Connect worker. Should not we introduce a barrier that would skip those two methods in case of `stop()/start()` sequence being in progress?

Reply:
Are you sure, is this not happening from within the polling loop?

Reply:
@gunnarmorling It is also indirectly called from `org.apache.kafka.connect.runtime.SourceTaskOffsetCommitter` which runs in its own executor thread

Reply:
Changed to use a reentrant read write lock to ensure exclusivity of critical sections now:

* start / stop / commit are all exclusive
* poll obtains a readlock to check whether we already restarted after a retry